### PR TITLE
Adding support for `UInt`-to-`char` conversion

### PR DESCRIPTION
**Pull request by viniciusfdasilva**
This pull request adds support for integer-to-char conversion, allowing the compiler to correctly handle character casting, implementing part of the issue #5922.

```carbon
import Core library "io";

fn Run() -> i32 {
	var i : i32 = 65;
	var ch: char = (i as char); // Support implemented!
	Core.PrintChar(ch); // Print 'A'
	return 0;
}
```

### CODE REVIEW

**Comment by danakj** [comment]
> This pull request adds support for i32-to-char conversion, enabling the compiler to correctly handle character casting from 32-bit integers.
> 
> ```v
> import Core library "io";
> 
> fn Run() -> i32 {
> 	var i : i32 = 65;
> 	var ch: char = (i as char); // Support implemented!
> 	Core.PrintChar(ch); // Print 'A'
> 	return 0;
> }
> ```

I don't see `i32` -> `char` listed in here as a builtin operation: https://github.com/carbon-language/carbon-lang/issues/5922#issuecomment-3161194912

Could you clarify where in the design or a proposal or leads issue this is from?

**Comment by danakj** [comment]
> > This pull request adds support for i32-to-char conversion, enabling the compiler to correctly handle character casting from 32-bit integers.
> > ```v
> > import Core library "io";
> > 
> > fn Run() -> i32 {
> > 	var i : i32 = 65;
> > 	var ch: char = (i as char); // Support implemented!
> > 	Core.PrintChar(ch); // Print 'A'
> > 	return 0;
> > }
> > ```
> 
> I don't see `i32` -> `char` listed in here as a builtin operation: [#5922 (comment)](https://github.com/carbon-language/carbon-lang/issues/5922#issuecomment-3161194912)
> 
> Could you clarify where in the design or a proposal or leads issue this is from?

Oh I see it:
> Explicit conversion to/from integers equivalent to what u8 supports.
> Continuing no implicit conversions.

Could you add a link to the PR description?

File: core/prelude/types/char.carbon [hunk 1]
```diff
@@ -27,3 +27,7 @@ impl CharLiteral() as ImplicitAs(Char) {
 impl CharLiteral() as As(Char) {
   fn Convert[self: Self]() -> Char = "char.convert_checked";
 }
+
+impl Int(32) as As(Char) {
```
Comment (danakj) [inline core/prelude/types/char.carbon]: Can we make this generic for all integer sizes, like it is for `u8`?

https://github.com/carbon-language/carbon-lang/blob/7be6538aec14ab5ef1bbedab45dcfbd15f96e62a/core/prelude/types/uint.carbon#L45

File: toolchain/check/eval.cpp [hunk 2]
```diff
@@ -867,14 +867,6 @@ static auto ResolveSpecificDeclForInst(EvalContext& eval_context,
...
           ResolveSpecificDeclForSpecificId(eval_context, interface.specific_id);
         }
-        for (const auto& constraint : info.extend_named_constraints) {
-          ResolveSpecificDeclForSpecificId(eval_context,
-                                           constraint.specific_id);
-        }
-        for (const auto& constraint : info.self_impls_named_constraints) {
-          ResolveSpecificDeclForSpecificId(eval_context,
-                                           constraint.specific_id);
-        }
```
Comment (danakj) [inline toolchain/check/eval.cpp:877]: This looks like a bad rebase?

File: toolchain/check/eval.cpp [hunk 3]
```diff
@@ -991,6 +983,7 @@ static auto PerformCheckedCharConvert(Context& context, SemIR::LocId loc_id,
                                       SemIR::InstId arg_id,
                                       SemIR::TypeId dest_type_id)
     -> SemIR::ConstantId {
+
```
Comment (danakj) [inline toolchain/check/eval.cpp]:
```suggestion
```

File: toolchain/check/eval.cpp [hunk 4]
```diff
@@ -1035,6 +1028,48 @@ static auto MakeFloatTypeResult(Context& context, SemIR::LocId loc_id,
   return MakeConstantResult(context, result, phase);
 }
 
+//// Performs a conversion from integer to character type.
+// Performs a conversion between character types, diagnosing if the value
+// doesn't fit in the destination type.
```
Comment (danakj) [inline toolchain/check/eval.cpp]: This looks like a copy paste that needs to be cleaned up?

File: toolchain/check/eval.cpp [hunk 5]
```diff
@@ -1035,6 +1028,48 @@ static auto MakeFloatTypeResult(Context& context, SemIR::LocId loc_id,
...
+static auto PerformCharConvert(Context& context, SemIR::LocId loc_id,
+                                      SemIR::InstId arg_id,
```
Comment (danakj) [inline toolchain/check/eval.cpp]: I suggest setting up pre-commit locally so things like formatting are fixed before uploading. https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md

File: toolchain/check/eval.cpp [hunk 6]
```diff
@@ -1035,6 +1028,48 @@ static auto MakeFloatTypeResult(Context& context, SemIR::LocId loc_id,
   return MakeConstantResult(context, result, phase);
 }
 
+//// Performs a conversion from integer to character type.
+// Performs a conversion between character types, diagnosing if the value
+// doesn't fit in the destination type.
+static auto PerformCharConvert(Context& context, SemIR::LocId loc_id,
```
Comment (danakj) [inline toolchain/check/eval.cpp]:
```suggestion
static auto PerformIntToCharConvert(Context& context, SemIR::LocId loc_id,
```

File: toolchain/check/eval.cpp [hunk 7]
```diff
@@ -1035,6 +1028,48 @@ static auto MakeFloatTypeResult(Context& context, SemIR::LocId loc_id,
...
+        SemIR::TypeId);
+    context.emitter().Emit(loc_id, NegativeIntInUnsignedType,
+                           {.type = arg.type_id, .value = arg_val},
+                           dest_type_id);
+  }
+  
+  int64_t value = arg_val.getSExtValue();
+
+  // Values over 0x80 require multiple code units in UTF-8.
+  if (value >= 0x80) {
```
Comment (danakj) [inline toolchain/check/eval.cpp]: This surprised me, could you point me to where the design says this should happen, rather than converting anything up to 255?

File: toolchain/sem_ir/builtin_function_kind.cpp [hunk 8]
```diff
@@ -373,6 +373,7 @@ constexpr BuiltinInfo PrintInt = {
 constexpr BuiltinInfo ReadChar = {"read.char",
                                   ValidateSignature<auto()->AnySizedInt>};
 
+
```
Comment (danakj) [inline toolchain/sem_ir/builtin_function_kind.cpp]:
```suggestion
```

File: toolchain/sem_ir/builtin_function_kind.cpp [hunk 9]
```diff
@@ -784,6 +789,7 @@ auto BuiltinFunctionKind::IsCompTimeOnly(const File& sem_ir,
       // Checked conversions are compile-time only.
       return true;
 
+    case IntConvertChar:
     case IntConvert:
```
Comment (danakj) [inline toolchain/sem_ir/builtin_function_kind.cpp]:
```suggestion
    case IntConvert:
    case IntConvertChar:
```
Prefixes sort first

File: toolchain/sem_ir/builtin_function_kind.def [hunk 10]
```diff
@@ -45,6 +45,7 @@ CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(MaybeUnformedMakeType)
 CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(CharConvertChecked)
 
 // Integer conversion.
+CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(IntConvertChar)
 CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(IntConvert)
```
Comment (danakj) [inline toolchain/sem_ir/builtin_function_kind.def]:
```suggestion
CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(IntConvert)
CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(IntConvertChar)
```

**Review by danakj — COMMENTED** [review]
Please add tests including both tests that now pass and tests that fail for bad values or types, and tests that pass/fail for making a "convert_char" builtin signature that is correct or malformed.

File: core/prelude/types/char.carbon [hunk 11]
```diff
@@ -27,3 +27,7 @@ impl CharLiteral() as ImplicitAs(Char) {
...
+  fn Convert[self: Self]() -> Char = "int.convert_char";  
+}
```
Comment (CarbonInfraBot) [inline core/prelude/types/char.carbon]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
  fn Convert[self: Self]() -> Char = "int.convert_char";
}

```

File: toolchain/sem_ir/builtin_function_kind.cpp [hunk 12]
```diff
@@ -411,6 +412,10 @@ constexpr BuiltinInfo CharConvertChecked = {
     "char.convert_checked",
     ValidateSignature<auto(CharLiteral)->CharCompatible>};
 
+// Converts from an integer type to a char-compatible type (u8/adapted Char).
+constexpr BuiltinInfo IntConvertChar = {"int.convert_char", 
+                                        ValidateSignature<auto(AnyInt)->CharCompatible>};
```
Comment (CarbonInfraBot) [inline toolchain/sem_ir/builtin_function_kind.cpp]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
constexpr BuiltinInfo IntConvertChar = {
    "int.convert_char", ValidateSignature<auto(AnyInt)->CharCompatible>};
```

File: toolchain/sem_ir/builtin_function_kind.cpp (hunk 8 above)
Comment (viniciusfdasilva) [reply toolchain/sem_ir/builtin_function_kind.cpp]: Thank you!

File: toolchain/sem_ir/builtin_function_kind.cpp (hunk 9 above)
Comment (viniciusfdasilva) [reply toolchain/sem_ir/builtin_function_kind.cpp]: Sure!

File: toolchain/check/eval.cpp (hunk 4 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp]: Yes!! I will remove this

File: toolchain/check/eval.cpp (hunk 2 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp:877]: This might be a branch update problem! 
I'll fix that as well! 
This code should be included.

File: toolchain/check/eval.cpp (hunk 7 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp]: Actually, it’s a restriction I initially included, but it may indeed not make sense to keep it.

File: core/prelude/types/char.carbon (hunk 1 above)
Comment (viniciusfdasilva) [reply core/prelude/types/char.carbon]: Good idea!

Comment (viniciusfdasilva) [reply core/prelude/types/char.carbon]: Hi @danakj 

I had issues with this kind of implementation, but I managed to get support working for i32, i8, and i64! However, I’ll work on adapting it to what you suggested in the future.

File: toolchain/check/eval.cpp (hunk 5 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp]: Thank you! I used pre-commit in my new submition :)

File: toolchain/check/eval.cpp (hunk 7 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp]: Yes! I removed this code snippet! I don’t think it makes sense to add this check.

File: toolchain/check/eval.cpp (hunk 2 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp:877]: I added this snippet again!

File: core/prelude/types/char.carbon (hunk 1 above)
Comment (danakj) [reply core/prelude/types/char.carbon]: Could you say more what issues you had? I think we should be able to write this as a single generic.

File: toolchain/check/eval.cpp [hunk 13]
```diff
@@ -1820,6 +1848,12 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
     }
 
     // Integer conversions.
+    case SemIR::BuiltinFunctionKind::IntConvertChar: {
+      if (phase != Phase::Concrete) {
+        return MakeConstantResult(context, call, phase);
+      }
+      return PerformIntToCharConvert(context, loc_id, arg_ids[0], call.type_id);
```
Comment (danakj) [inline toolchain/check/eval.cpp]: What if we used the pre-existing PerformIntConvert() here? Would that work? Do we want the behaviour to differ in any way? I don't recall seeing that we do.

File: core/prelude/types/char.carbon (hunk 1 above)
Comment (viniciusfdasilva) [reply core/prelude/types/char.carbon]: Hi @danakj  I adapted to this code:

```carbon
impl forall [From:! IntLiteral(), To:! CharLiteral()] From as As(To) { 
  fn Convert[self: Self]() -> Char = "int.convert_char";
}
```

And, I got the following error: 

```text
/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/types/char.carbon:33:55: error: cannot implicitly convert non-type value of type `Core.IntLiteral` to `type`
impl forall [From:! IntLiteral(), To:! CharLiteral()] From as As(To) { 
                                                      ^~~~
/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/types/char.carbon:33:55: note: type `Core.IntLiteral` does not implement interface `ImplicitAs(type)`
impl forall [From:! IntLiteral(), To:! CharLiteral()] From as As(To) { 
                                                      ^~~~

/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/types/char.carbon:33:63: error: cannot implicitly convert non-type value of type `Core.CharLiteral` to `type`
impl forall [From:! IntLiteral(), To:! CharLiteral()] From as As(To) { 
                                                              ^~~~~~
/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/types/char.carbon:33:63: note: type `Core.CharLiteral` does not implement interface `ImplicitAs(type)`
impl forall [From:! IntLiteral(), To:! CharLiteral()] From as As(To) { 
                                                              ^~~~~~
/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/types/char.carbon:7:1: in import
/home/silva/.cache/bazel/_bazel_silva/d567bf4ac99d39d82d9ac8f68f7be289/execroot/_main/bazel-out/k8-fastbuild/bin/toolchain/install/prefix_root/bin/../lib/carbon/../../lib/carbon/core/prelude/operators/as.carbon:13:14: note: initializing generic parameter `Dest` declared here
interface As(Dest:! type) {
```


This error does not seem too difficult to fix; I just need a bit of time to take a look at it. We can merge this PR with the changes as they are and I can work on this improvement later. What do you think?

File: toolchain/check/eval.cpp (hunk 13 above)
Comment (viniciusfdasilva) [reply toolchain/check/eval.cpp]: Hi @danakj, from what I understand, PerformIntConvert() is called after converting x to an integer value. But in this case, we want to convert from an integer to a character, and the name PerformIntConvert() wouldn’t make sense in this context. What do you think? 

That's why I created one called `PerformIntToCharConvert()`

File: core/prelude/types/char.carbon (hunk 1 above)
Comment (danakj) [reply core/prelude/types/char.carbon]: Did you mean `UInt(From)` and `UInt(To)`? The impl needs to be using types, and IntLiteral is a value of type IntLiteralType, not of type `type`.

Edit: Rather, `UInt(From)`  and `Char` in this case.

Comment (danakj) [reply core/prelude/types/char.carbon]: > We can merge this PR with the changes as they are and I can work on this improvement later. What do you think?

This is the primary point of the PR, I think we should just figure it out.

**Review by danakj — COMMENTED** [review]
(moving this out of my review queue for now)

File: core/prelude/types/char.carbon (hunk 1 above)
Comment (viniciusfdasilva) [reply core/prelude/types/char.carbon]: @danakj Solved by commit [4b6ea2e](https://github.com/carbon-language/carbon-lang/pull/6425/commits/4b6ea2e2f3dbe4b07c568c3d0371b5a51ac0c65c)

File: toolchain/check/eval.cpp (hunk 13 above)
Comment (danakj) [reply toolchain/check/eval.cpp]: Did you have any thoughts about this?

**Review by danakj — COMMENTED** [review]
(telling github I am waiting for a reply on https://github.com/carbon-language/carbon-lang/pull/6425#discussion_r2665292300, it didn't capture that as a "review")

**Comment by viniciusfdasilva** [comment]
> (telling github I am waiting for a reply on [#6425 (comment)](https://github.com/carbon-language/carbon-lang/pull/6425#discussion_r2665292300), it didn't capture that as a "review")

Hi @danakj, from what I understand, `PerformIntConvert()` is called after converting x to an integer value. But in this case, we want to convert from an integer to a character, and the name `PerformIntConvert()` wouldn’t make sense in this context.

That's why I created one called `PerformIntToCharConvert()`

 What do you think?

**Review by viniciusfdasilva — COMMENTED** [review]
Hi @danakj, from what I understand, PerformIntConvert() is called after converting x to an integer value. But in this case, we want to convert from an integer to a character, and the name PerformIntConvert() wouldn’t make sense in this context.

That's why I created one called PerformIntToCharConvert()

What do you think?

**Review by viniciusfdasilva — COMMENTED** [review]
Hi @danakj, from what I understand, PerformIntConvert() is called after converting x to an integer value. But in this case, we want to convert from an integer to a character, and the name PerformIntConvert() wouldn’t make sense in this context.

That's why I created one called PerformIntToCharConvert()

What do you think?

**Comment by danakj** [comment]
> Hi @danakj, from what I understand, PerformIntConvert() is called after converting x to an integer value.

I don't think so? It's called in the same place, just below, and does the same kind of work. It converts an integer to another integer. Char is a type of integer (we call `GetIntTypeInfo` in the same way).

> That's why I created one called PerformIntToCharConvert()

Sure, if we want different behaviour for converting to a char than converting to a `u8` but do we?

**Review by danakj — COMMENTED** [review]
(replied elsewhere)

**Review by viniciusfdasilva — COMMENTED** [review]
> I don't think so? It's called in the same place, just below, and does the same kind of work. It converts an integer to another integer. Char is a type of integer (we call GetIntTypeInfo in the same way).

Yes, but I think that if we put the logic implemented in this pull request into `PerformIntConvert`, we may end up with code that is quite large and confusing. I believe that creating `PerformIntToCharConvert` helps with readability. But if necessary, I can move the logic into `PerformIntConvert`.


> Sure, if we want different behaviour for converting to a char than converting to a `u8` but do we?

I'm not sure if I understood the question, but this pull request currently implements the explicit conversion from `uint` to `char`, not from `char` to `uint`.

```carbon
// Implemented on this PR!
var i : u8 = 64;
var ch: char = (i as char);
```
```carbon
// Not implemented on this PR!
var i : char = 'A';
var ch: u8 = (i as u8);
```

**Comment by danakj** [comment]
> > I don't think so? It's called in the same place, just below, and does the same kind of work. It converts an integer to another integer. Char is a type of integer (we call GetIntTypeInfo in the same way).
> 
> Yes, but I think that if we put the logic implemented in this pull request into `PerformIntConvert`, we may end up with code that is quite large and confusing. I believe that creating `PerformIntToCharConvert` helps with readability. But if necessary, I can move the logic into `PerformIntConvert`.
> 
> > Sure, if we want different behaviour for converting to a char than converting to a `u8` but do we?
> 
> I'm not sure if I understood the question, but this pull request currently implements the explicit conversion from `uint` to `char`, not from `char` to `uint`.
> 
> ```v
> // Implemented on this PR!
> var i : u8 = 64;
> var ch: char = (i as char);
> ```
> 
> ```v
> // Not implemented on this PR!
> var i : char = 'A';
> var ch: u8 = (i as u8);
> ```

Yes, I understand, but my question does not seem to be coming across. I am wondering about replacing the call to `PerformIntToCharConvert` with a call to `PerformIntConvert`, and not making any changes to `PerformIntConvert`. So it would work the exact same to do u32-to-u8 and u32-to-char. Do we want u32-to-char to be different?

**Review by viniciusfdasilva — COMMENTED** [review]
> Yes, I understand, but my question does not seem to be coming across. I am wondering about replacing the call to `PerformIntToCharConvert` with a call to `PerformIntConvert`, and not making any changes to `PerformIntConvert`. So it would work the exact same to do u32-to-u8 and u32-to-char. Do we want u32-to-char to be different?

 You'e right! Replacing PerformIntToCharConvert with PerformIntConvert worked and has the same behavior!

File: toolchain/check/testdata/builtins/int/convert.carbon [hunk 14]
```diff
@@ -297,3 +297,47 @@ let convert_not_constant: i16 = IntLiteralToInt16(not_constant);
...
+
+  // UTF-8 single-byte values (128–255)
+  Test(Int32ToChar(0x80)) as Expect(0x80 as char);
+  Test(Int32ToChar(0x90)) as Expect(0x90 as char);
+  Test(Int32ToChar(0xA0)) as Expect(0xA0 as char);
+  Test(Int32ToChar(0xC0)) as Expect(0xC0 as char);
+  Test(Int32ToChar(0xE0)) as Expect(0xE0 as char);
+  Test(Int32ToChar(0xFF)) as Expect(0xFF as char);
+
+  // Negative values → must error (same style as sign_extend test)
```
Comment (danakj) [inline toolchain/check/testdata/builtins/int/convert.carbon]: I think the tests will need to be updated now?

**Review by viniciusfdasilva — COMMENTED** [review]
> I think the tests will need to be updated now?

 Done!

File: toolchain/check/testdata/builtins/int/convert.carbon [hunk 15]
```diff
@@ -297,3 +302,61 @@ let convert_not_constant: i16 = IntLiteralToInt16(not_constant);
...
+  // UTF-8 single-byte values (128–255)
+  Test(UInt8ToChar(0x80)) as Expect(0x80 as char);
+  Test(UInt8ToChar(0x90)) as Expect(0x90 as char);
+  Test(UInt8ToChar(0xA0)) as Expect(0xA0 as char);
+  Test(UInt8ToChar(0xC0)) as Expect(0xC0 as char);
+  Test(UInt8ToChar(0xE0)) as Expect(0xE0 as char);
+  Test(UInt8ToChar(0xFF)) as Expect(0xFF as char);
+
+  // for u32
+  // Basic ASCII range
```
Comment (CarbonInfraBot) [inline toolchain/check/testdata/builtins/int/convert.carbon]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
  // Basic ASCII range
```

File: toolchain/check/testdata/builtins/int/convert.carbon [hunk 16]
```diff
@@ -297,3 +302,61 @@ let convert_not_constant: i16 = IntLiteralToInt16(not_constant);
...
+  // UTF-8 single-byte values (128–255)
+  Test(UInt32ToChar(0x80)) as Expect(0x80 as char);
+  Test(UInt32ToChar(0x90)) as Expect(0x90 as char);
+  Test(UInt32ToChar(0xA0)) as Expect(0xA0 as char);
+  Test(UInt32ToChar(0xC0)) as Expect(0xC0 as char);
+  Test(UInt32ToChar(0xE0)) as Expect(0xE0 as char);
+  Test(UInt32ToChar(0xFF)) as Expect(0xFF as char);
+
+  // for u64
+  // Basic ASCII range
```
Comment (CarbonInfraBot) [inline toolchain/check/testdata/builtins/int/convert.carbon]: **[diff]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>
```suggestion
  // Basic ASCII range
```

File: toolchain/check/testdata/builtins/int/convert.carbon [hunk 17]
```diff
@@ -304,40 +309,54 @@ library "[[@TEST_NAME]]";
 import library "int_ops";
```
Comment (danakj) [inline toolchain/check/testdata/builtins/int/convert.carbon]: The tests all come before the semir lines (the stuff like `// CHECK:STDOUT:   %Int32ToInt16.call: ...`, so these should move earlier in the file. I don't think it's going to pass before that. Did you run autoupdate? That should regenerate those lines too, and what you upload has to match what CI produces when it runs autoupdate.

File: toolchain/check/testdata/builtins/int/convert.carbon [hunk 18]
```diff
@@ -304,40 +309,54 @@ library "[[@TEST_NAME]]";
...
 fn F() {
+  // for u8
   // Basic ASCII range
-  Test(Int32ToChar(0)) as Expect(0 as char);
-  Test(Int32ToChar(10)) as Expect(10 as char);
-  Test(Int32ToChar(32)) as Expect(32 as char);
-  Test(Int32ToChar(65)) as Expect(65 as char);
-  Test(Int32ToChar(97)) as Expect(97 as char);
-  Test(Int32ToChar(0x7F)) as Expect(0x7F as char);
+  Test(UInt8ToChar(0)) as Expect(0 as char);
```
Comment (danakj) [inline toolchain/check/testdata/builtins/int/convert.carbon]: There are both too many and too few cases in here, I think. We want to test boundary conditions, but we don't need to test so much in between them. Just copying from the tests above, we can test for u8:
- 0x00
- 0x80
- 0xFF

For u16 (missing atm)
- 0x0000
- 0xFF12 comes out as 0x12
- 0xFFFF comes out as 0xFF

For u32
- 0x0000_0000
- 0x8000_0012 comes out as 0x12
- 0xFFFF_FF12 comes out as 0x12
- 0xFFFF_FFFF comes out as 0xFF

For u64
- 0x0
- 0x8000_0000_0000_0012 comes out as 0x12
- 0xFFFF_FFFF_FFFF_FF12 comes out as 0x12
- 0xFFFF_FFFF_FFFF_FFFF comes out as 0xFF

Sound right?

**Review by viniciusfdasilva — COMMENTED** [review]
> The tests all come before the semir lines (the stuff like // CHECK:STDOUT: %Int32ToInt16.call: ..., so these should move earlier in the file. I don't think it's going to pass before that. Did you run autoupdate? That should regenerate those lines too, and what you upload has to match what CI produces when it runs autoupdate.

No, I didn’t run autoupdate! Please, how can I find the documentation to run it?

**Comment by chandlerc** [comment]
> > The tests all come before the semir lines (the stuff like // CHECK:STDOUT: %Int32ToInt16.call: ..., so these should move earlier in the file. I don't think it's going to pass before that. Did you run autoupdate? That should regenerate those lines too, and what you upload has to match what CI produces when it runs autoupdate.
> 
> No, I didn’t run autoupdate! Please, how can I find the documentation to run it?

https://docs.carbon-lang.dev/toolchain/docs/adding_features.html#updating-tests

**Review by danakj — COMMENTED** [review]
(clearing review queue while waiting for test changes)

**Review by viniciusfdasilva — COMMENTED** [review]
I think now It's ready! :)

**Comment by danakj** [comment]
> I think now It's ready! :)

Thanks for all the iterating on this, I can see how dramatically the PR simplified as we worked on it. Double checking the tests, but LGTM

**Review by danakj — APPROVED** [review]
